android/keyple-plugin/android-cone2/src/main/java/org/eclipse/keyple/plugin/android/cone2/Cone2PluginImpl.java

```diff
@@ -111,65 +118,114 @@ protected SeReader fetchNativeReader(String name) throws KeypleReaderException {
-                        @Override
-                        public void onInstanceAvailable(Reader reader) {
-                            initNativeReaders();
-                        }
-
-                        @Override
-                        public void onError(int error) {
 
+                    @Override
+                    public void onSuccess(CpcResult.RESULT result) {
+                        if (readers != null) {
+                            for (SeReader reader : readers) {
+                                readersNames.add(reader.getName());
+                            }
                         }
-                    });
-                } else {
-                    isReaderPoweredOn.set(false);
-                    Cone2AskReader.clearInstance();
-                    notifyObservers(new PluginEvent(PLUGIN_NAME, readersNames, PluginEvent.EventType.READER_DISCONNECTED));
-                }
-            }
-
-            @Override
-            public void onError(Throwable e) {
+                        isReaderPoweredOn.set(true);
+                        Cone2AskReader.getInstance(context, new Cone2AskReader.ReaderListener() {
+                            @Override
+                            public void onInstanceAvailable(Reader reader) {
+                                initNativeReaders();
+                            }
+
+                            @Override
+                            public void onError(int error) {
+                                LOG.error("Error trying to get CpcAsk.Reader instance");
+                            }
+                        });
+                    }
 
-            }
-        });
+                    @Override
+                    public void onError(Throwable e) {
+                        LOG.error("Error trying to power on the reader");
+                    }
+                });
     }
 
-    @Override
-    protected SortedSet<String> fetchNativeReadersNames() throws KeypleReaderException {
-        return readersNames;
+    /**
+     * Powers off reader and stops card discovery.
+     * @param context Context
+     */
+    private void powerOff(Context context) {
+        for (SeReader reader : readers) {
+            if (reader.getName().compareTo(Cone2ContactlessReader.READER_NAME) == 0) {
+                final Cone2ContactlessReaderImpl cone2Reader = (Cone2ContactlessReaderImpl) reader;
+                cone2Reader.stopWaitForCard();
+
+                // This completable monitors the waitForCardPresent method. while it is running, it
+                // is waiting for it to finish before powezring off the reader.
+                // If the
+                Completable.create(new CompletableOnSubscribe() {
+                    @Override
+                    public void subscribe(CompletableEmitter emitter) {
+                        while (!cone2Reader.waitForCardPresentStopped()) {
+                            SystemClock.sleep(10);
```

Comment:
Est-ce que tu veux bien regarder si tu peux remplacer l'atomicBoolean qui est en dessous par une ConditionVariable : https://baptiste-wicht.com/posts/2010/09/java-concurrency-part-5-monitors-locks-and-conditions.html 

Je n'aime pas trop les sleeps

Reply:
Oui je vais regarder, en même temps il ne faut pas être dogmatique, le sleep ici n'a pour seul vocation que d'empêcher un while d'écraser le CPU. ON n'est pas dans le cas d'un sleep mis pour caler un problème de délai.

Reply:
Oui, c'est vrai.
Si ça ne fonctionne pas avec une condition variable on gardera cette implémentation qui a le mérite de fonctionner. Je pense toutefois que c'est intéressant de trouver une manière de se passer de sleep.

Reply:
Oui je vais simplement partir sur un Lock. Comme j'ai déjà utilisé pour la gestion, partagée du lecteur ASK. C'est beaucoup plus propre, tu as raison.

Reply:
Voilà c'est fait, pour commentaires.